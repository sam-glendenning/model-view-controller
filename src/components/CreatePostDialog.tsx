import {
  Dialog,
  DialogTitle,
  DialogContent,
  Box,
  TextField,
  DialogActions,
  Button,
} from '@mui/material';
import React, { useCallback } from 'react';
import { useCreatePostController } from '@/controllers/useCreatePostController';

interface CreatePostDialogProps {
  isOpen: boolean;
  onClose: () => void;
  onSuccess?: (message: string) => void;
  onError?: (message: string) => void;
}

export const CreatePostDialog: React.FC<CreatePostDialogProps> = ({
  isOpen,
  onClose,
  onSuccess,
  onError,
}) => {
  const {
    formData,
    isSubmitDisabled,
    handleClose,
    updateTitle,
    updateBody,
    updateUserId,
    submitForm,
  } = useCreatePostController({
    onClose,
    onSuccess,
    onError,
  });

  const handleSubmit = useCallback(async () => {
    await submitForm();
  }, [submitForm]);

  return (
    <Dialog open={isOpen} onClose={handleClose} maxWidth="sm" fullWidth>
      <DialogTitle>Create New Post</DialogTitle>
      <DialogContent>
        <Box sx={{ pt: 1 }}>
          <TextField
            autoFocus
            margin="dense"
            label="Title"
            type="text"
            fullWidth
            variant="outlined"
            value={formData.title}
            onChange={e => updateTitle(e.target.value)}
            sx={{ mb: 2 }}
          />
          <TextField
            margin="dense"
            label="Content"
            type="text"
            fullWidth
            multiline
            rows={4}
            variant="outlined"
            value={formData.body}
            onChange={e => updateBody(e.target.value)}
            sx={{ mb: 2 }}
          />
          <TextField
            margin="dense"
            label="User ID"
            type="number"
            fullWidth
            variant="outlined"
            value={formData.userId}
            onChange={e => updateUserId(parseInt(e.target.value) ?? 1)}
          />
        </Box>
      </DialogContent>
      <DialogActions>
        <Button onClick={handleClose}>Cancel</Button>
        <Button
          onClick={handleSubmit}
          variant="contained"
          disabled={isSubmitDisabled}
        >
          Create
        </Button>
      </DialogActions>
    </Dialog>
  );
};
