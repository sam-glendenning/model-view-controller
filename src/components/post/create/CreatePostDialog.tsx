import React from 'react';
import {
  Dialog,
  DialogTitle,
  DialogContent,
  Box,
  TextField,
  DialogActions,
  Button,
} from '@mui/material';
import type { Post } from '@/types';

interface CreatePostDialogProps {
  // Dialog state
  open: boolean;
  onClose: () => void;

  // Form data
  postData: Post;
  onTitleChange: (title: string) => void;
  onBodyChange: (body: string) => void;
  onUserIdChange: (userId: number) => void;

  // Actions
  onSubmit: () => Promise<void>;
  isCreating: boolean;
}

export const CreatePostDialog: React.FC<CreatePostDialogProps> = ({
  open,
  onClose,
  postData,
  onTitleChange,
  onBodyChange,
  onUserIdChange,
  onSubmit,
  isCreating,
}) => {
  return (
    <Dialog open={open} onClose={onClose} maxWidth="sm" fullWidth>
      <DialogTitle>Create New Post</DialogTitle>
      <DialogContent>
        <Box sx={{ pt: 1 }}>
          <TextField
            autoFocus
            margin="dense"
            label="Title"
            type="text"
            fullWidth
            variant="outlined"
            value={postData.title}
            onChange={e => onTitleChange(e.target.value)}
            sx={{ mb: 2 }}
          />
          <TextField
            margin="dense"
            label="Content"
            type="text"
            fullWidth
            multiline
            rows={4}
            variant="outlined"
            value={postData.body}
            onChange={e => onBodyChange(e.target.value)}
            sx={{ mb: 2 }}
          />
          <TextField
            margin="dense"
            label="User ID"
            type="number"
            fullWidth
            variant="outlined"
            value={postData.userId}
            onChange={e => onUserIdChange(parseInt(e.target.value) ?? 1)}
          />
        </Box>
      </DialogContent>
      <DialogActions>
        <Button onClick={onClose}>Cancel</Button>
        <Button onClick={onSubmit} variant="contained" disabled={isCreating}>
          Create
        </Button>
      </DialogActions>
    </Dialog>
  );
};
