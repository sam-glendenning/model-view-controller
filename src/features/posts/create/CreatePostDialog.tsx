import React from 'react';
import {
  Box,
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
  TextField,
} from '@mui/material';
import type { Post } from '@/shared/types';

interface CreatePostDialogProps {
  // Dialog state
  open: boolean;
  onClose: () => void;

  // Form data
  postData: Omit<Post, 'id'>;
  onTitleChange: (title: string) => void;
  onBodyChange: (body: string) => void;
  onUserIdChange: (userId: number) => void;

  // Actions
  onSubmit: () => Promise<void>;
  isCreating: boolean;
}

export const CreatePostDialog: React.FC<CreatePostDialogProps> = ({
  open,
  onClose,
  postData,
  onTitleChange,
  onBodyChange,
  onUserIdChange,
  onSubmit,
  isCreating,
}) => {
  return (
    <Dialog open={open} onClose={onClose} maxWidth="sm" fullWidth>
      <DialogTitle>Create New Post</DialogTitle>
      <DialogContent>
        <Box sx={{ pt: 1 }}>
          <TextField
            autoFocus
            margin="dense"
            label="Title"
            type="text"
            fullWidth
            variant="outlined"
            value={postData.title}
            onChange={e => {
              onTitleChange(e.target.value);
            }}
            sx={{ mb: 2 }}
          />
          <TextField
            margin="dense"
            label="Content"
            type="text"
            fullWidth
            multiline
            rows={4}
            variant="outlined"
            value={postData.body}
            onChange={e => {
              onBodyChange(e.target.value);
            }}
            sx={{ mb: 2 }}
          />
          <TextField
            margin="dense"
            label="User ID"
            type="number"
            fullWidth
            variant="outlined"
            value={postData.userId}
            onChange={e => {
              const parsed = parseInt(e.target.value);
              onUserIdChange(Number.isNaN(parsed) ? 1 : parsed);
            }}
          />
        </Box>
      </DialogContent>
      <DialogActions>
        <Button onClick={onClose}>Cancel</Button>
        <Button
          onClick={() => void onSubmit()}
          variant="contained"
          disabled={isCreating}
        >
          Create
        </Button>
      </DialogActions>
    </Dialog>
  );
};
