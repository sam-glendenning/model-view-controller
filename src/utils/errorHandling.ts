import { ApiError } from '@/types';
import { AxiosError } from 'axios';

/**
 * Type guard to check if an error is an AxiosError
 * @param error - The error to check
 * @returns True if the error is an AxiosError
 */
export const isAxiosError = (error: unknown): error is AxiosError => {
  return error instanceof Error && 'isAxiosError' in error && error.isAxiosError === true;
};

/**
 * Transforms any error into a standardized ApiError
 * @param error - The error to transform
 * @returns Standardized ApiError object
 */
export const transformToApiError = (error: unknown): ApiError => {
  // Handle AxiosError
  if (isAxiosError(error)) {
    const responseData = error.response?.data as any;
    return {
      message: responseData?.message || error.message || 'Network request failed',
      status: error.response?.status,
      code: error.code,
    };
  }
  
  // Handle regular Error
  if (error instanceof Error) {
    return {
      message: error.message,
    };
  }
  
  // Handle string errors
  if (typeof error === 'string') {
    return {
      message: error,
    };
  }
  
  // Handle unknown errors
  return {
    message: 'An unknown error occurred',
  };
};

/**
 * Gets a user-friendly error message from an error
 * @param error - The error to get the message from
 * @returns User-friendly error message
 */
export const getErrorMessage = (error: unknown): string => {
  const apiError = transformToApiError(error);
  
  // Map common HTTP status codes to user-friendly messages
  if (apiError.status) {
    switch (apiError.status) {
      case 400:
        return 'Invalid request. Please check your input and try again.';
      case 401:
        return 'You are not authorized to perform this action.';
      case 403:
        return 'Access denied. You do not have permission to perform this action.';
      case 404:
        return 'The requested resource was not found.';
      case 422:
        return 'Invalid data provided. Please check your input.';
      case 429:
        return 'Too many requests. Please wait a moment and try again.';
      case 500:
        return 'Server error. Please try again later.';
      case 502:
      case 503:
      case 504:
        return 'Service unavailable. Please try again later.';
      default:
        return apiError.message;
    }
  }
  
  return apiError.message;
};
