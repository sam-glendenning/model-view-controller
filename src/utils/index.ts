export * from './validation';
export * from './errorHandling';
