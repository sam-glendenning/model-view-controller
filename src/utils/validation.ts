import { ValidationResult, PostFormData } from '@/types';

/**
 * Validates post form data
 * @param formData - The form data to validate
 * @returns ValidationResult with isValid status and any errors
 */
export const validatePostForm = (formData: PostFormData): ValidationResult => {
  const errors: string[] = [];
  
  // Title validation
  if (!formData.title || formData.title.trim().length === 0) {
    errors.push('Title is required');
  } else if (formData.title.trim().length < 3) {
    errors.push('Title must be at least 3 characters long');
  } else if (formData.title.trim().length > 100) {
    errors.push('Title must be less than 100 characters');
  }
  
  // Body validation
  if (!formData.body || formData.body.trim().length === 0) {
    errors.push('Content is required');
  } else if (formData.body.trim().length < 10) {
    errors.push('Content must be at least 10 characters long');
  } else if (formData.body.trim().length > 1000) {
    errors.push('Content must be less than 1000 characters');
  }
  
  // User ID validation
  if (!formData.userId || formData.userId <= 0) {
    errors.push('Valid user ID is required');
  } else if (!Number.isInteger(formData.userId)) {
    errors.push('User ID must be a whole number');
  }
  
  return {
    isValid: errors.length === 0,
    errors,
  };
};

/**
 * Type guard to check if form data is valid for creating a post
 * @param formData - The form data to check
 * @returns True if the form data is valid for creating a post
 */
export const isValidCreatePostForm = (formData: PostFormData): formData is PostFormData => {
  return validatePostForm(formData).isValid;
};

/**
 * Type guard to check if a value is a valid post ID
 * @param id - The ID to check
 * @returns True if the ID is a valid post ID
 */
export const isValidPostId = (id: unknown): id is number => {
  return typeof id === 'number' && id > 0 && Number.isInteger(id);
};

/**
 * Type guard to check if a value is a valid user ID
 * @param id - The ID to check
 * @returns True if the ID is a valid user ID
 */
export const isValidUserId = (id: unknown): id is number => {
  return typeof id === 'number' && id > 0 && Number.isInteger(id);
};
