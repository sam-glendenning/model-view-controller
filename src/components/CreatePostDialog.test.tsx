import { render } from '@/test/utils';
import { CreatePostDialog } from './CreatePostDialog';

const mockOnClose = jest.fn();
const mockOnSuccess = jest.fn();
const mockOnError = jest.fn();

describe('CreatePostDialog', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders snapshot correctly', () => {
    const { baseElement } = render(
      <CreatePostDialog
        isOpen={true}
        onClose={mockOnClose}
        onSuccess={mockOnSuccess}
        onError={mockOnError}
      />,
    );

    expect(baseElement).toMatchSnapshot();
  });
});
