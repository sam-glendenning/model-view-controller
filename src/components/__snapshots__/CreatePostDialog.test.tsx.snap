// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreatePostDialog renders snapshot correctly 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiDialog-root MuiModal-root css-zw3mfo-MuiModal-root-MuiDialog-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root MuiModal-backdrop css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
      role="presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        aria-labelledby="mui-1"
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiDialog-paperFullWidth css-tlc64q-MuiPaper-root-MuiDialog-paper"
        role="dialog"
      >
        <h2
          class="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-bdhsul-MuiTypography-root-MuiDialogTitle-root"
          id="mui-1"
        >
          Create New Post
        </h2>
        <div
          class="MuiDialogContent-root css-ypiqx9-MuiDialogContent-root"
        >
          <div
            class="MuiBox-root css-zoser8"
          >
            <div
              class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth MuiTextField-root css-3tho05-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-focused MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="true"
                for="mui-3"
                id="mui-3-label"
              >
                Title
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth Mui-focused MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id="mui-3"
                  type="text"
                  value=""
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-r01yff-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-sl37lx-MuiNotchedOutlined-root"
                  >
                    <span>
                      Title
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <div
              class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth MuiTextField-root css-3tho05-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="false"
                for="mui-4"
                id="mui-4-label"
              >
                Content
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline css-8ewcdo-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <textarea
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                  id="mui-4"
                  rows="4"
                  style="height: 0px; overflow: hidden;"
                />
                <textarea
                  aria-hidden="true"
                  class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
                  readonly=""
                  style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding-top: 0px; padding-bottom: 0px; width: 100%;"
                  tabindex="-1"
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-r01yff-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-njhac4-MuiNotchedOutlined-root"
                  >
                    <span>
                      Content
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
            <div
              class="MuiFormControl-root MuiFormControl-marginDense MuiFormControl-fullWidth MuiTextField-root css-1z10yd4-MuiFormControl-root-MuiTextField-root"
            >
              <label
                class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
                data-shrink="true"
                for="mui-5"
                id="mui-5-label"
              >
                User ID
              </label>
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
              >
                <input
                  aria-invalid="false"
                  class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                  id="mui-5"
                  type="number"
                  value="1"
                />
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-r01yff-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-sl37lx-MuiNotchedOutlined-root"
                  >
                    <span>
                      User ID
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiDialogActions-root MuiDialogActions-spacing css-knqc4i-MuiDialogActions-root"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1e6y48t-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Cancel
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-sghohy-MuiButtonBase-root-MuiButton-root"
            disabled=""
            tabindex="-1"
            type="button"
          >
            Create
          </button>
        </div>
      </div>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
